src/test/extensions/provisioningHelper.spec.ts

```diff
@@ -0,0 +1,210 @@
+import * as nock from "nock";
+import * as api from "../../api";
+import * as provisioningHelper from "../../extensions/provisioningHelper";
+import * as extensionsApi from "../../extensions/extensionsApi";
+import { expect } from "chai";
+import { FirebaseError } from "../../error";
+
+const TEST_INSTANCES_RESPONSE = {};
+const PROJECT_ID = "test-project";
+const SPEC_WITH_STORAGE_AND_AUTH = {
+  apis: [
+    {
+      apiName: "storage-component.googleapis.com",
+    },
+    {
+      apiName: "identitytoolkit.googleapis.com",
+    },
+  ] as extensionsApi.Api[],
+  resources: [] as extensionsApi.Resource[],
+} as extensionsApi.ExtensionSpec;
+
+const FIREDATA_AUTH_ACTIVATED_RESPONSE = {
+  activation: [
+    {
+      service: "FIREBASE_AUTH",
+    },
+  ],
+};
+
+const FIREBASE_STORAGE_DEFAULT_BUCKET_LINKED_RESPONSE = {
+  buckets: [
+    {
+      name: `projects/12345/bucket/${PROJECT_ID}.appspot.com`,
+    },
+  ],
+};
+
+describe.only("provisioningHelper", () => {
+  afterEach(() => {
+    nock.cleanAll();
+  });
+
+  describe("getUsedProducts", () => {
+    it("returns empty array when nothing is used", () => {
+      expect(
+        provisioningHelper.getUsedProducts({
+          apis: [
+            {
+              apiName: "unrelated.googleapis.com",
+            },
+          ] as extensionsApi.Api[],
+          roles: [
+            {
+              role: "unrelated.role",
+            },
+          ] as extensionsApi.Role[],
+          resources: [
+            {
+              propertiesYaml:
+                "availableMemoryMb: 1024\neventTrigger:\n  eventType: providers/unrelates.service/eventTypes/something.do\n  resource: projects/_/buckets/${param:IMG_BUCKET}\nlocation: ${param:LOCATION}\nruntime: nodejs10\n",
+            },
+          ] as extensionsApi.Resource[],
+        } as extensionsApi.ExtensionSpec)
+      ).to.be.empty;
+    });
+    it("returns STORAGE when Storage API is used", () => {
+      expect(
+        provisioningHelper.getUsedProducts({
+          apis: [
```

Comment:
There's a lot of repetivie boilerplate code in this file for building these test objects. Consider using a beforeEach() block to deduplicate some of this:
```
let testSpec
beforeEach(() => {
 testSpec = {
          roles: [ ] as extensionsApi.Role[],
          resources: [] as extensionsApi.Resource[],
        } as extensionsApi.ExtensionSpec)
})
```
Then, in each it block, just assign what you need:
```
it("returns STORAGE when Storage Role is used", () => {
  testSpec.roles.push( { role: "storage.object.admin" });

  expect(provisioningHelper.getUsedProducts(testSpec)).to.be...
}
```

Reply:
Done.